pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -164,13 +165,67 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
 	if t.pollsToArchive <= 0 || t.persister == nil {
 		return
 	}
-	key := fmt.Sprintf("knownFiles%d", t.archiveIndex)
-	if err := checkpoint.SaveKey(context.Background(), t.persister, metadata.Get(), key); err != nil {
+	if err := t.updateArchive(t.archiveIndex, metadata); err != nil {
 		t.set.Logger.Error("error faced while saving to the archive", zap.Error(err))
 	}
 	t.archiveIndex = (t.archiveIndex + 1) % t.pollsToArchive // increment the index
 }
 
+func (t *fileTracker) readArchive(index int) (*fileset.Fileset[*reader.Metadata], error) {
+	// readArchive loads data from the archive for a given index and returns a fileset.Filset.
+	key := fmt.Sprintf("knownFiles%d", index)
+	metadata, err := checkpoint.LoadKey(context.Background(), t.persister, key)
+	if err != nil {
+		return nil, err
+	}
+	f := fileset.New[*reader.Metadata](len(metadata))
+	f.Add(metadata...)
+	return f, nil
+}
+
+func (t *fileTracker) updateArchive(index int, rmds *fileset.Fileset[*reader.Metadata]) error {
+	// updateArchive saves data to the archive for a given index and returns an error, if encountered.
+	key := fmt.Sprintf("knownFiles%d", index)
+	return checkpoint.SaveKey(context.Background(), t.persister, rmds.Get(), key)
+}
+
+func (t *fileTracker) SyncOffsets() {
+	// SyncOffsets goes through all new (unmatched) readers and updates the metadata, if found on archive.
```

Comment:
This wasn't my understanding of how the archive would work. We shouldn't have to reconcile multiple copies of the same metadata and worry about syncing. It's too complicated. 

We should use almost the exact same patterns as when searching `knownFiles`. The only difference is that in order to reduce inefficient read/writes to storage, we should search 1 fileset for N fingerprints before moving on to the next fileset. (As opposed to searching all filesets for 1 fingerprint before moving on to the next fingerprint.)

Searching the archive should look something like:

```go
func (t *fileTracker) FindFiles(fps []fingerprint.Fingerprint) []reader.Metadata {
    matchedFPs := make([]reader.Metadata)
    for i := mostRecentIndex; i != mostRecentIndex; i = (i + 1) % pollsToArchive {
        fs := loadFileset(i)  // the entire fileset at most once poll
        var modified bool
        for _, fp := range fps {
            if matchedFP := fs.Get(fp) { // removes fp if matched
               matchedFPs = append(matchedFPs, matchedFP)
               modified = true
            }
        }
        if modified {
            saveFileset(i, fs) // overwrite the entire fileset at most once poll
        }            
    return matchedFPs
}
```

This way filesets in the archive are kept up to date in near-real time. The only time when they could be out of data is before the function returns. It also ensures we are minimizing interactions with the archive. (One thing I didn't include in my pseudocode is early exit if all fingerprints have been found, but we should do that too.)

Reply:
From an high level, this is what I did. But syncing/reconcile made it complex. I agree with you.